rosco-manifests/src/main/java/com/netflix/spinnaker/rosco/manifests/helmfile/HelmfileTemplateUtils.java

```diff
@@ -0,0 +1,144 @@
+import com.netflix.spinnaker.kork.exceptions.SpinnakerException;
+import com.netflix.spinnaker.kork.retrofit.exceptions.SpinnakerHttpException;
+import com.netflix.spinnaker.rosco.jobs.BakeRecipe;
+import com.netflix.spinnaker.rosco.manifests.ArtifactDownloader;
+import com.netflix.spinnaker.rosco.manifests.BakeManifestEnvironment;
+import com.netflix.spinnaker.rosco.manifests.HelmBakeTemplateUtils;
+import com.netflix.spinnaker.rosco.manifests.config.RoscoHelmConfigurationProperties;
+import com.netflix.spinnaker.rosco.manifests.config.RoscoHelmfileConfigurationProperties;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class HelmfileTemplateUtils extends HelmBakeTemplateUtils<HelmfileBakeManifestRequest> {
+  private final RoscoHelmfileConfigurationProperties helmfileConfigurationProperties;
+  private final RoscoHelmConfigurationProperties helmConfigurationProperties =
+      new RoscoHelmConfigurationProperties();
+
+  public HelmfileTemplateUtils(
+      ArtifactDownloader artifactDownloader,
+      RoscoHelmfileConfigurationProperties helmfileConfigurationProperties) {
+    super(artifactDownloader);
+    this.helmfileConfigurationProperties = helmfileConfigurationProperties;
+  }
+
+  public BakeRecipe buildBakeRecipe(
+      BakeManifestEnvironment env, HelmfileBakeManifestRequest request) throws IOException {
+    BakeRecipe result = new BakeRecipe();
+    result.setName(request.getOutputName());
+
+    Path helmfileFilePath;
+    List<Path> valuePaths;
+
+    List<Artifact> inputArtifacts = request.getInputArtifacts();
+    if (inputArtifacts == null || inputArtifacts.isEmpty()) {
+      throw new IllegalArgumentException("At least one input artifact must be provided to bake");
+    }
+
+    log.info("helmfileFilePath: '{}'", request.getHelmfileFilePath());
+    Artifact helmfileTemplateArtifact = inputArtifacts.get(0);
+    String artifactType = Optional.ofNullable(helmfileTemplateArtifact.getType()).orElse("");
+    if ("git/repo".equals(artifactType)) {
+      env.downloadArtifactTarballAndExtract(super.getArtifactDownloader(), helmfileTemplateArtifact);
+
+      // If there's no helmfile path specified, assume it lives in the root of
+      // the git/repo artifact.
+      helmfileFilePath =
+          env.resolvePath(Optional.ofNullable(request.getHelmfileFilePath()).orElse(""));
+    } else {
+      try {
+        helmfileFilePath = downloadArtifactToTmpFile(env, helmfileTemplateArtifact);
+      } catch (SpinnakerHttpException e) {
+        throw new SpinnakerHttpException(fetchFailureMessage("template", e), e);
+      } catch (IOException | SpinnakerException e) {
+        throw new IllegalStateException(fetchFailureMessage("template", e), e);
+      }
+    }
```

Comment:
seems like this code is duplicated

Reply:
Yes it is duplicated but it is not easy to reuse. Please if you can recommend any ways to reuse this code I would appreciate it. Thanks

Reply:
You should be able to extract duplication either to a utilities class or similar in `-core` type projects.  Possibly rewriting some of the inputs/outputs to be more "generic" on downloading to a tmp file.  Note I'm PRETTY sure kustomize uses some of this same logic, so a generic "extract artifact to folder" utility that passes the resulting FS path.  

I'd also argue for a chroot type setup where each one runs in isolation or you can hit some interesting concurrency issues (which we kinda already get with tmp file support.. but... I've been seeing more edge cases lately).  Don't think THIS ask is doable, but moving to a "core" lib should be.

Reply:
Thanks @jasonmcintosh to provide some ideas. I'm going to check definitely whether the "core" lib way is a doable thing. I would like to avoid to introduce heavy refactors but certainly at this point it is too early to say that. I'll get back here once I have a more concrete idea.

Reply:
can we move this method(`buildBakeRecipe`) to `HelmBakeTemplateUtils` class and add an abstract method for the command(`buildCommand`, `getCommand`)?

Reply:
@nemesisOsorio the whole `buildBakeRecipe` depends by `request` POJO which differs from `Helm` and `Helmfile` so moving `buildBakeRecipe` into `HelmBakeTemplateUtils` won't solve the problem.

Reply:
With the last changes, everything now should result more dry and favourite the reuse.